src/client_impl.cc

```diff
@@ -20,64 +20,82 @@ using ::istio::mixer::v1::CheckRequest;
 using ::istio::mixer::v1::CheckResponse;
 using ::istio::mixer::v1::ReportRequest;
 using ::istio::mixer::v1::ReportResponse;
-using ::istio::mixer::v1::QuotaRequest;
-using ::istio::mixer::v1::QuotaResponse;
 using ::google::protobuf::util::Status;
 using ::google::protobuf::util::error::Code;
 
 namespace istio {
 namespace mixer_client {
 
 MixerClientImpl::MixerClientImpl(const MixerClientOptions &options)
-    : options_(options) {
+    : options_(options), deduplication_id_(0) {
   check_cache_ =
       std::unique_ptr<CheckCache>(new CheckCache(options.check_options));
   report_batch_ = std::unique_ptr<ReportBatch>(
       new ReportBatch(options.report_options, options_.report_transport,
                       options.timer_create_func, converter_));
-  quota_cache_ = std::unique_ptr<QuotaCache>(new QuotaCache(
-      options.quota_options, options_.quota_transport, converter_));
+  quota_cache_ =
+      std::unique_ptr<QuotaCache>(new QuotaCache(options.quota_options));
 }
 
 MixerClientImpl::~MixerClientImpl() {}
 
 void MixerClientImpl::Check(const Attributes &attributes, DoneFunc on_done) {
-  std::string signature;
-  Status status =
-      check_cache_->Check(attributes, system_clock::now(), &signature);
-  if (status.error_code() != Code::NOT_FOUND) {
-    on_done(status);
+  std::unique_ptr<CheckCache::CacheResult> check_result(
+      new CheckCache::CacheResult);
+  check_cache_->Check(attributes, check_result.get());
+  if (check_result->IsCacheHit() && !check_result->status().ok()) {
+    on_done(check_result->status());
     return;
   }
 
+  std::unique_ptr<QuotaCache::CacheResult> quota_result(
+      new QuotaCache::CacheResult);
+  // Only use quota cache if Check is using cache with OK status.
+  // Otherwise, a remote Check call may be rejected, but quota amounts were
+  // substracted from quota cache already.
+  quota_cache_->Quota(attributes, check_result->IsCacheHit(),
+                      quota_result.get());
+
   CheckRequest request;
+  bool quota_call = quota_result->BuildRequest(&request);
+  if (check_result->IsCacheHit() && quota_result->IsCacheHit()) {
+    on_done(quota_result->status());
+    on_done = nullptr;
+    if (!quota_call) {
+      return;
+    }
+  }
+
   converter_.Convert(attributes, request.mutable_attributes());
   request.set_global_word_count(converter_.global_word_count());
+  request.set_deduplication_id(std::to_string(deduplication_id_++));
```

Comment:
The dedup id needs to be a true UUID. Otherwise, requests from different proxies may get dedupped against one another.